src/goto-instrument/contracts/contracts.cpp

```diff
@@ -287,7 +290,8 @@ void code_contractst::check_apply_loop_contracts(
       converter.goto_convert(old_decreases_assignment, generated_code, mode);
     }
 
-    goto_function.body.destructive_insert(loop_body, generated_code);
+    goto_function.body.destructive_insert(
+      loop_body, add_pragma_disable_assigns_check(generated_code));
```

Comment:
shouldn't the decreases clause's initial value be computed before the loop's guard expression is evaluated ?

on this program  

```
while (i < max++)
loop_assigns(i, max)
loop_decreases(max-i)
{
  i++;
}
```
instrumenting like this refutes the loop decreases clause
```
havoc(i, max);
decr_begin = max-i; 
max += 1;
if(i < max)
{
  i++;
  decr_end = max-i;
  assert(decr_end < decr_begin); // falsified
}
```

but instrumenting like this would tell you it holds 

```
havoc(i, max);
max += 1;
if(i < max)
{
  decr_begin = max-i; 
  i++;
  decr_end = max-i;
  assert(decr_end < decr_begin); // proved
}
```

Moreover the goto code for this program we can see that the backjump is before the guard's side effects computation:

```
int main()
{
  unsigned int max;
  unsigned int i = 0;
  while (i < max++)
  {
    i++;
  }
  return 0;
}

```

```
main /* main */
        // 0 file test.c line 6 function main
        DECL main::1::max : signedbv[32]
        // 1 file test.c line 7 function main
        DECL main::1::i : signedbv[32]
        // 2 file test.c line 7 function main
        ASSIGN main::1::i := 0
        // 3 file test.c line 8 function main
     1: DECL main::$tmp::tmp_post : signedbv[32]
        // 4 file test.c line 8 function main
        ASSIGN main::$tmp::tmp_post := main::1::max
        // 5 file test.c line 8 function main
        ASSIGN main::1::max := main::1::max + 1
        // 6 file test.c line 8 function main
        IF ¬(main::1::i < main::$tmp::tmp_post) THEN GOTO 2
        // 7 file test.c line 10 function main
        ASSIGN main::1::i := main::1::i + 1
        // 8 file test.c line 8 function main
        GOTO 1
        // 9 file test.c line 8 function main
     2: SKIP
        // 10 file test.c line 12 function main
        SET RETURN VALUE 0
        // 11 file test.c line 12 function main
        DEAD main::$tmp::tmp_post
        // 12 file test.c line 12 function main
        DEAD main::1::i
        // 13 file test.c line 12 function main
        DEAD main::1::max
        // 14 file test.c line 12 function main
        GOTO 3
        // 15 file test.c line 13 function main
     3: END_FUNCTION
```

Reply:
We had discussed this yesterday and I think this should be a separate PR.

I also think we would find more bugs to squash if we want to handle loop guards that have side effects. That was out of scope until now.